transform/map_mutex.go

```diff
@@ -0,0 +1,38 @@
+package transform
+
+import (
+	"sync"
```

Comment:
não é importante para mudança, mas queria deixar outra opção para fazer mapas com mutex, poderia usar isso https://pkg.go.dev/sync

Reply:
Você diz o `sync.Map`? Eu pensei nele, mas como ele usa `interface{}` eu fiquei confuso se não seria trabalho demais fazer o casting para só guradar um nada (podia ser um `struct{}{}` hehehe…). E ainda perderia a segurança de tipos…
O que acha?

Reply:
não sei se entendi bem, eu nunca tive que fazer casting dos tipos, ele sempre funcionou bem, eu vi um benchmark que era menos performático que usar o sync e a gente lidar com o mutex, mas pela simplicidade do codigo as vezes compensa(não to dizendo que é o caso), eu falei dele para te mostrar a opção, não sabia se conhecia ele

Reply:
Preciso estudar mais o `sync.Map` acho! Tentei implementar aqui começando com algo assim:

```go
package transform

import (
	"sync"
)

type mapMutex sync.Map

func (m *mapMutex) Load(k interface{}) (interface{}, bool) { return m.Load(k) }
func (m *mapMutex) Store(k interface{}, v interface{})     { m.Store(k, v) }
func (m *mapMutex) Delete(k interface{})                   { m.Delete(k) }

func (m *mapMutex) lock(k string) {
	for {
		_, exists := m.Load(k)
		if !exists {
			m.Store(k, struct{}{})
			return
		}
	}
}

func (m *mapMutex) unlock(k string) {
	m.Delete(k)
}

func newMapMutex() mapMutex {
	return mapMutex{}
}
```

Deu _stack overflow_ hehehe…

```
runtime: goroutine stack exceeds 1000000000-byte limit
runtime: sp=0xc022400390 stack=[0xc022400000, 0xc042400000]
fatal error: stack overflow
```

Reply:
Implementei algo melhor no 3939cee — achei bem melhor! Obrigado pela dica ; )